iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces2.cpp

```diff
@@ -260,47 +92,1009 @@ static LogicalResult declareTargetOps(TargetOptions targetOptions,
+    // I/O to an external/unknowable buffer (invocation capture/escape).
+    EXTERNAL_BUFFER = 1u << 2,
+    // Push constant (index) non-buffer type.
+    PUSH_CONSTANT = 1u << 3,
+  };
+
+  // Operand index in the flattened dispatch function.
+  int operandIndex = 0;
+
+  // Bitfield of usage throughout the program.
+  // Entry points dispatched from many places may have differing usage.
+  unsigned usage = Usage::UNKNOWN;
+
+  // Storage buffer and ranges pulled from it if this has CONSTANT_BUFFER usage
+  // and all uses are of the same storage buffer. Unset if multiple storage
+  // buffers are used.
+  SymbolRefAttr constantBuffer;
+  SmallVector<Attribute> constantRanges;
+
+  void dump() {
+    if (usage & DerivedOperandInfo::Usage::PUSH_CONSTANT) {
+      llvm::dbgs() << "  PUSH_CONSTANT\n";
+    }
+    if (usage & DerivedOperandInfo::Usage::CONSTANT_BUFFER) {
+      if (constantBuffer) {
+        llvm::dbgs() << "  CONSTANT_BUFFER " << constantBuffer << "\n";
+        for (auto &range : constantRanges) {
+          llvm::dbgs() << "   RANGE " << range << "\n";
+        }
+      } else {
+        llvm::dbgs() << "  CONSTANT_BUFFER (NON-UNIFORM STORAGE)\n";
+      }
+    }
+    if (usage & DerivedOperandInfo::Usage::TRANSIENT_BUFFER) {
+      llvm::dbgs() << "  TRANSIENT_BUFFER\n";
+    }
+    if (usage & DerivedOperandInfo::Usage::EXTERNAL_BUFFER) {
+      llvm::dbgs() << "  EXTERNAL_BUFFER\n";
+    }
+  }
+};
+
+// Derives information about each dispatch region operand based on all of the
+// |dispatchOps| which dispatch it. Returns a vector mapped 1:1 with the
+// |entryFuncOp| operand list of flow.dispatch.tensor types.
+static SmallVector<DerivedOperandInfo> deriveOperandInfo(
+    FuncOp entryFuncOp, ArrayRef<IREE::Flow::DispatchOp> dispatchOps) {
+  // Mapping of flattened dispatch region operands to binding information.
+  SmallVector<DerivedOperandInfo> operandInfos;
+  operandInfos.resize(entryFuncOp.getNumArguments());
+
+  SmallVector<OperandResultRef> regionToDispatchMap;
+  regionToDispatchMap.resize(entryFuncOp.getNumArguments());
+  for (int i = 0; i < entryFuncOp.getNumArguments(); ++i) {
+    regionToDispatchMap[i] = mapRegionOperandToDispatchValue(entryFuncOp, i);
+  }
+
+  for (auto dispatchOp : dispatchOps) {
+    for (unsigned operandIndex = 0;
+         operandIndex < entryFuncOp.getNumArguments(); ++operandIndex) {
+      auto &operandInfo = operandInfos[operandIndex];
+      operandInfo.operandIndex = operandIndex;
+      auto ref = regionToDispatchMap[operandIndex];
+      if (ref.isOperand()) {
+        auto operand = dispatchOp.operands()[ref.index()];
+        auto operandType = operand.getType();
+        if (operandType.isIntOrIndexOrFloat()) {
+          operandInfo.usage = DerivedOperandInfo::Usage::PUSH_CONSTANT;
+        } else if (operandType.isa<TensorType>()) {
+          if (auto subspanOp = dyn_cast_or_null<IREE::HAL::ConstantSubspanOp>(
+                  operand.getDefiningOp())) {
```

Comment:
nit: Some deep nesting happening in here. Could invert in a few places with some `continue;` statements.

(e.g. if ref is not operand, label as external_buffer and `continue`)

Reply:
unnested the result but the next change I'm making changes this all so I'll leave the operandType switch as-is